### PR TITLE
[FEATURE] Improve comments in grammar

### LINKED ISSUE #1705: [BUG] Comments are not allowed in all places
We wrote comments to be mixed with other comments but not to be used *in* commands, like this:

```
choices is ['batu', 'kertas', 'gunting']
print 'Pilihlah'
print 'batu, kertas, gunting'
userc is ask 'Pilihan anda?'
print 'Pilihan anda: ' userc
computerc is choices at random
print 'Pilihan komputer: ' computerc
if userc is computerc:
    print 'SERI'
elif userc is 'batu' and computerc is 'kertas':
    print 'Komputer menang' #prints something <------ does not work now
else:
    print 'Komputer menang'
```

I guess rather than adding comments in the grammar as we do now, it would be easier to write a little preprocessor that removes all code on a line after the #. We already have preprocessing code in place, so that is not all too hard and we don't care about the comments anyway.

Maybe a fun issue for @racertop(when you are done with the input stuff, of course!)?

### PR DESCRIPTION
**Pull request by TiBiBa**
**Description**
In this PR we fix an issue where the comment was not allowed in all places. We add an `%ignore` rule to make sure everything after the `#` is always ignored. We also add the `#` to the `text` rule to make sure something is a text until a newline **or** a #.

**Fixes**
This PR fixes #1705

**How to test**
This is still a WiP, I'm currently figuring out how to write code test cases and verify the rule correctness.

### CODE REVIEW

**Comment by Felienne** [comment]
Ah nice, Lark has support for comments, I did not know! Great find. I guess we also want to add a few tests? And maybe @jpelay wants to review? (sorry that this is now being picked up, I know you also wanted to work on it, but there are enough issues open haha)

**Comment by TiBiBa** [comment]
> Ah nice, Lark has support for comments, I did not know! Great find. I guess we also want to add a few tests? And maybe @jpelay wants to review? (sorry that this is now being picked up, I know you also wanted to work on it, but there are enough issues open haha)

Yes apparently there is an `%ignore` option, this way we don't have to parse for the comments as they are automatically ignored. I'm still trying to figure out a lot about Lark and getting familiar with the syntax. For now this seems to simplify the current way we handle comments (but inline comments are still a WiP....). Will keep working on this issue in this branch until I figure it out!

**Comment by TiBiBa** [comment]
> Cool thanks @TiBiBa! If you need help, let me know!

It seems to work, but it also seems too easy a fix to be working. Can you verify?

**Comment by TiBiBa** [comment]
The current idea is expanding on the already existing `text` rule:

- Everything is `text` until a newline
   - Expand with -> everything is a `text` until a newline **or** a #

**Comment by TiBiBa** [comment]
Still have to figure some stuff out as this is working:
Schermafbeelding 2022-03-23 om 14 23 16

And this isn't:
Schermafbeelding 2022-03-23 om 14 23 25

**Comment by TiBiBa** [comment]
The last case is fixed as well by keeping the current comment rule as well:
Schermafbeelding 2022-03-23 om 14 30 12

**Comment by TiBiBa** [comment]
Can verify that the example code in the issue is indeed working as expected:
Schermafbeelding 2022-03-23 om 14 42 03

**Comment by jpelay** [comment]
Hi, Timon! Such a and smart great solution,

**Comment by jpelay** [comment]
Hi,  Timon, this such an elegant and smart solution :) Your test are failing because at level 4 we don't use the grammar merger due to weird interactions in the precedence of the rules, so we have to redefine the command rule again. You just gotta delete the command rule from there, something like this:

`command:print | ask | turtle | assign_list | add | remove | sleep | error_ask_no_quotes| assign | error_print_no_quotes | error_invalid`


Now, the adventures are failing because many programs that are `start_code` consists only of one comment, and programs that consists of only comments raise `EmptyProgramException` because since comments are ignored, there isn't a rule that the `IsValid` filter can grab, and therefore detects it as an empty program.

**Comment by jpelay** [comment]
I took the freedom to make the tests pass, hope it's ok! These were just a few changes:

1. Changed the level 4 and level 5 grammar to remove the comment rule, and in the level 5 I allowed empty programs to be possible (as in the previews levels) this leads to,
2. Since now we allow empty programs it's no longer necessary the emptyProgramException, so I commented,the test on levels 1 that relate to this (if we check empty programs, programs that consitsts of comments alone will not pass, therefore many adventures programs fail) I think these tests are not needed anymore, but Im commenting just in case
3.  I also added the # to some text rules from level 4 on.

**Comment by jpelay** [comment]
Worth checking! comment after repeat fails!
image

**Comment by jpelay** [comment]
Again, sorry I jumped in! Hope it's ok, Timon! I was reviewing the changes and got a bit carried away :sweat_smile: There's a test, however, that stills fails, and it regards the numeric translations, not sure what's happening there yet

**Comment by TiBiBa** [comment]
> Again, sorry I jumped in! Hope it's ok, Timon! I was reviewing the changes and got a bit carried away There's a test, however, that stills fails, and it regards the numeric translations, not sure what's happening there yet

No problem, thanks a lot for helping out! Will take a look at the last failing test later today.

**Comment by TiBiBa** [comment]
Just checking out the last failing test, but not sure why it is failing. Verified the correctness in the code editor itself:

Schermafbeelding 2022-03-24 om 09 57 13

**Comment by TiBiBa** [comment]
Another think to keep in mind:

We should remove the "good job" success message in case of an empty program warning. As we no longer return an actual error we still have the (relic code) "you have no output" warning that catches the empty program.

**Comment by Felienne** [comment]
> Another think to keep in mind:
> 
> We should remove the "good job" success message in case of an empty program warning. As we no longer return an actual error we still have the (relic code) "you have no output" warning that catches the empty program.

Ah good catch, maybe make a separate issue for that? If you want I can also look into why the Chinese test is failing (you can open an issue for me)

**Comment by Felienne** [comment]
And thanks @jpelay for your extensive help on this!!

**Comment by TiBiBa** [comment]
I just fixed the last failing test, but by simply commenting it out in the .yaml file. Not really nice, but other than that it should all work now.

**Comment by thjazi39** [comment]
I have the code snippets ready to add syntax highlighting for comments.
I'll add them on this branch, or I'll do that once the merge #2264 is done (to reduce the risk of conflict?)

**Comment by Felienne** [comment]
I would sa add them here because thematically it goes together

**Comment by thjazi39** [comment]
It's done, I added syntax highlighting for the comments:
image

**Comment by Felienne** [comment]
I think this is good to go! Maybe @jpelay wants to do the final approval?

**Comment by jpelay** [comment]
> I think this is good to go! Maybe @jpelay wants to do the final approval? If you are happy, simply leave an approving review and it will merge automatically

Yes, I think everything is good! Although after the latest merge some tests fail that didn't fail before :< So I don't know if we can still merge. I'm checking them right now, but can't pinpoint the issue just yet.

**Comment by jpelay** [comment]
Not sure why the test fails, but everything works fine on my local in the editor so Im merging.

**Review by jpelay — APPROVED** [review]

**Comment by Felienne** [comment]
The issue seems to be with the changes to Missing Command. This test now gives a MissingCommand exception instead of an InvalidArgumentTypeException at level 5. 

```
  def test_add_to_list_with_string_var_gives_error(self):
    code = textwrap.dedent("""\
    color is yellow 
    colors is green, red, blue
    add colors to color""")

    self.multi_level_tester(
      max_level=11,
      code=code,
      exception=hedy.exceptions.InvalidArgumentTypeException
    )
```

Seems to be a task for @boryanagoncharenko?

**Comment by Felienne** [comment]
Converting to draft until @boryanagoncharenko figures out the ambig issues

File: grammars/level5-Additions.lark [hunk 1]
```diff
@@ -1,4 +1,6 @@
-program: _EOL* (command | error_invalid)* (_SPACE)* (_EOL+ (command | error_invalid) (_SPACE)*)* _EOL* //lines may end on spaces and might be separated by many newlines
+program: _empty_program | _non_empty_program
+_empty_program: (_EOL | _SPACE)*
```
Comment (Felienne) [inline grammars/level5-Additions.lark:2]: Smart fix @boryanagoncharenko!!

**Review by Felienne — APPROVED** [review]

**Comment by boryanagoncharenko** [comment]
The level 5 grammar is adapted and the failing zh test is re-added, so there are no intermittent test failures. Changing the PR status from Draft -> Ready For Review.

File: grammars/level5-Additions.lark (hunk 1 above)
Comment (boryanagoncharenko) [reply grammars/level5-Additions.lark:2]: It is the only fix that actually solved the problem. I thought I understood why there is an ambiguity at level 6, but as I dived deeper I figured the ambiguity should already kick in at level 5. But it did not, which made me wonder whether I understood what's going wrong in the first place... So, I had to take this alternative approach. Grammar ambiguities are hard.

**Comment by Felienne** [comment]
What a team effort @jpelay, @boryanagoncharenko and @TiBiBa <3

**Pull request merged**